File: python/src/vaas/resource_owner_password_grant_authenticator.py
"""Verdict-as-a-Service

:mod:`vaas` is a Python library for the VaaS-API."""
from .vaas_errors import VaasAuthenticationError
from authlib.integrations.httpx_client import AsyncOAuth2Client

class ResourceOwnerPasswordGrantAuthenticator:
    """Tracing interface for Vaas"""

    def __init__(self, client_id, user_name, password, token_endpoint, verify=True):
        self.client_id = client_id
        self.user_name = user_name
        self.password = password
        self.token_endpoint = token_endpoint
        self.verify = verify

    async def get_token(self):
        async with AsyncOAuth2Client(
                self.client_id, verify=self.verify
        ) as client:
            try:
                token = \
                (await client.fetch_token(self.token_endpoint, username=self.user_name, password=self.password, grant_type="password"))[
                    "access_token"]
                return token
            except Exception as e:
                raise VaasAuthenticationError(e)

File: ruby/lib/vaas/resource_owner_password_grant_authenticator.rb
require 'json'
require 'async'
require 'async/http/internet'

module VAAS
  class ResourceOwnerPasswordGrantAuthenticator

    attr_accessor :client_id, :token_endpoint, :token, :username, :password

    def initialize(client_id, username, password, token_endpoint = 'https://account.gdata.de/realms/vaas-production/protocol/openid-connect/token')
      @client_id = client_id
      @username = username
      @password = password
      @token_endpoint = token_endpoint
    end

    def get_token
      Async do
        client = Async::HTTP::Internet.new

        header = [['content-type', 'application/x-www-form-urlencoded']]
        body = ["grant_type=password&client_id=#{client_id}&username=#{username}&password=#{password}"]

        response = client.post(token_endpoint, header, body)
        self.token = JSON.parse(response.read)['access_token']
      rescue => e
        raise VaasAuthenticationError, e
      ensure
        client&.close
      end
      token
    end
  end
end
